server/src/tree-sitter/analyzer.ts

```diff
@@ -374,73 +375,112 @@ export default class Analyzer {
     return tree.rootNode.descendantForPosition({ row: line, column })
   }
 
-  // The following functions are from SymbolScanner.ts
-  private extendsFile (filePath: string): void {
-    logger.debug(`extendsFile file: ${filePath}`)
-
+  /**
+   *
+   * @param uri
+   * @param document Main purpose of this param is to avoid re-reading the file from disk and re-parsing the tree when the file is opened for the first time since the same process will happen in the analyze() before calling this function.
+   */
+  public sourceIncludeFiles (uri: string, document?: { td: TextDocument, tree: Parser.Tree }): void {
+    if (this.parser === undefined) {
+      logger.error('[Analyzer] The analyzer is not initialized with a parser')
+      return
+    }
+    const filePath = uri.replace('file://', '')
+    logger.debug(`[Analyzer] Sourcing file: ${filePath}`)
     try {
-      const data: Buffer = fs.readFileSync(filePath)
-      const file: string[] = data.toString().split(/\r?\n/g)
+      let textDocument: TextDocument
+      let parsedTree: Parser.Tree
+      if (document !== undefined) {
+        textDocument = document.td
+        parsedTree = document.tree
+      } else {
+        const analyzedDocument = this.uriToAnalyzedDocument[uri]
+        if (analyzedDocument === undefined) {
+          textDocument = TextDocument.create(
+            uri,
+            'bitbake',
+            0,
+            fs.readFileSync(uri.replace('file://', ''), 'utf8')
+          )
+          parsedTree = this.parser.parse(textDocument.getText())
+          // Store it in analyzedDocument just like what analyze() does to avoid re-reading the file from disk and re-parsing the tree when editing on the same file
+          this.uriToAnalyzedDocument[uri] = {
+            document: textDocument,
+            globalDeclarations: getGlobalDeclarations({ tree: parsedTree, uri }), // TODO: Aovid doing this operation as it is not needed during the sourcing process
+            embeddedRegions: getEmbeddedRegionsFromNode(parsedTree, uri), // TODO: Avoid doing this operation as it is not needed during the sourcing process
```

Comment:
Are these TODOs addressed in this PR, or left for later?

Reply:
For the 2 TODOs here:
The 1st TODO has been removed as you can see in the later commit because the `globalDeclarations` is used.
The 2nd TODO is not addressed yet since it affects the work by @idillon-sfl . My goal here is to make `embeddedRegions` optional.